src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -78,80 +81,120 @@ public void onRefreshSettings(Settings settings) {
         this.bufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.buffer_size", ByteSizeValue.parseBytesSizeValue("64k")).bytes(); // Not really interesting, updated by IndexingMemoryController...
-        this.transientBufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.transient_buffer_size", ByteSizeValue.parseBytesSizeValue("8k")).bytes();
 
-        indexSettingsService.addListener(applySettings);
+        if (indexSettingsService != null) {
+            indexSettingsService.addListener(applySettings);
+        }
+        recoverFromFiles();
+        // new that we know which files are there, create a new current one.
+        current = createTranslogFile();
     }
 
-    public FsTranslog(ShardId shardId, @IndexSettings Settings indexSettings, Path location) throws IOException {
-        super(shardId, indexSettings);
-        this.indexSettingsService = null;
-        this.locations = new Path[]{location};
-        Files.createDirectories(location);
-        this.bigArrays = BigArrays.NON_RECYCLING_INSTANCE;
+    /** recover all translog files found on disk */
+    private void recoverFromFiles() throws IOException {
+        try (ReleasableLock _ = writeLock.acquire()) {
+            ArrayList<FsChannelImmutableReader> foundTranslogs = new ArrayList<>();
+            for (Path location : locations) {
+                try (DirectoryStream<Path> stream = Files.newDirectoryStream(location, TRANSLOG_FILE_PREFIX + "[0-9]*")) {
+                    for (Path file : stream) {
+                        final long id = parseIdFromFileName(file);
+                        if (id < 0) {
+                            // failed to extract id. logged in parseIdFromFileName
+                            continue;
+                        }
+                        idGenerator = Math.max(idGenerator, id + 1);
+                        final ChannelReference raf = new InternalChannelReference(id, location.resolve(getPath(id)), StandardOpenOption.READ);
+                        foundTranslogs.add(new FsChannelImmutableReader(id, raf, raf.channel().size(), FsChannelReader.UNKNOWN));
+                        logger.debug("found local translog with id [{}]", id);
+                    }
+                }
+            }
+            CollectionUtil.timSort(foundTranslogs, new Comparator<FsChannelReader>() {
+                @Override
+                public int compare(FsChannelReader o1, FsChannelReader o2) {
+                    return Long.compare(o1.translogId(), o2.translogId());
+                }
+            });
+            uncomittedTranslogs.addAll(foundTranslogs);
+        }
+    }
 
-        this.type = FsTranslogFile.Type.fromString(indexSettings.get("index.translog.fs.type", FsTranslogFile.Type.BUFFERED.name()));
-        this.bufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.buffer_size", ByteSizeValue.parseBytesSizeValue("64k")).bytes();
+    protected long parseIdFromFileName(Path translogFile) {
```

Comment:
This method can be public, static, and take a String instead of a Path, which would allow us to unit test it. It will require moving the logging up one level higher though, but still nice to be able to unit test.